File: automation-audit-tool/scoring.py
"""
Core scoring engine. All business logic lives here — no Streamlit imports.
"""

from config import (
    WEIGHTS,
    TIME_COST_OPTIONS,
    ERROR_RATE_OPTIONS,
    DEPENDENCY_RISK_OPTIONS,
    TIER_CONFIG,
    TIER_AUTOMATE_NOW,
    TIER_AUTOMATE_SOON,
    TIER_EVALUATE,
    TIER_LOW_PRIORITY,
    TIER_DO_NOT_AUTOMATE,
    TIER_CONDITIONAL,
    JUDGMENT_OVERRIDE_THRESHOLD,
)


def normalize_time(label: str) -> float:
    return float(TIME_COST_OPTIONS.get(label, 0))


def normalize_error(label: str) -> float:
    return float(ERROR_RATE_OPTIONS.get(label, 0))


def normalize_dependency(label: str) -> float:
    return float(DEPENDENCY_RISK_OPTIONS.get(label, 0))


def score_to_tier(score: float) -> str:
    for tier in [TIER_AUTOMATE_NOW, TIER_AUTOMATE_SOON, TIER_EVALUATE, TIER_LOW_PRIORITY]:
        if score >= TIER_CONFIG[tier]["min"]:
            return tier
    return TIER_LOW_PRIORITY


def calculate_priority_score(workflow: dict) -> dict:
    """
    workflow keys:
      name, description, time_cost, error_rate,
      dependency_risk, data_quality, judgment_pct
    Returns:
      score (float 0-100), tier (str), flag (str|None), breakdown (dict)
    """
    time_score       = normalize_time(workflow["time_cost"])
    error_score      = normalize_error(workflow["error_rate"])
    judgment_pct     = float(workflow["judgment_pct"])
    judgment_inverse = 100.0 - judgment_pct
    dep_score        = normalize_dependency(workflow["dependency_risk"])

    base_score = (
        time_score       * WEIGHTS["time_cost"] +
        error_score      * WEIGHTS["error_rate"] +
        judgment_inverse * WEIGHTS["judgment_inverse"] +
        dep_score        * WEIGHTS["dependency_risk"]
    )
    base_score = round(base_score, 1)

    breakdown = {
        "time_score":        time_score,
        "error_score":       error_score,
        "judgment_inverse":  judgment_inverse,
        "dependency_score":  dep_score,
    }

    # Data quality flag overrides tier (but not score)
    if workflow["data_quality"].startswith("Poor"):
        return {
            "score":     base_score,
            "tier":      TIER_CONDITIONAL,
            "flag":      "Fix data quality first — automate once inputs are structured",
            "breakdown": breakdown,
        }

    # Judgment override
    if judgment_pct > JUDGMENT_OVERRIDE_THRESHOLD:
        return {
            "score":     base_score,
            "tier":      TIER_DO_NOT_AUTOMATE,
            "flag":      "Human judgment is the value — automation would degrade quality",
            "breakdown": breakdown,
        }

    return {
        "score":     base_score,
        "tier":      score_to_tier(base_score),
        "flag":      None,
        "breakdown": breakdown,
    }


def calculate_overall_readiness(scored_workflows: list[dict]) -> dict:
    """
    Aggregate score across all workflows. Excludes Do Not Automate and
    Conditional workflows from the average (they don't represent actionable
    automation potential).
    """
    actionable = [
        w for w in scored_workflows
        if w["result"]["tier"] not in (TIER_DO_NOT_AUTOMATE, TIER_CONDITIONAL)
    ]

    if not actionable:
        avg = 0.0
    else:
        avg = sum(w["result"]["score"] for w in actionable) / len(actionable)

    counts = {tier: 0 for tier in TIER_CONFIG}
    for w in scored_workflows:
        tier = w["result"]["tier"]
        counts[tier] = counts.get(tier, 0) + 1

    return {
        "overall_score":   round(avg, 1),
        "tier_counts":     counts,
        "automate_now":    counts.get(TIER_AUTOMATE_NOW, 0),
        "do_not_automate": counts.get(TIER_DO_NOT_AUTOMATE, 0),
        "conditional":     counts.get(TIER_CONDITIONAL, 0),
        "actionable_count": len(actionable),
        "total_count":      len(scored_workflows),
    }


def rank_workflows(scored_workflows: list[dict]) -> list[dict]:
    """Sort workflows by score descending, with special tiers last."""
    special = [TIER_DO_NOT_AUTOMATE, TIER_CONDITIONAL]
    actionable = [w for w in scored_workflows if w["result"]["tier"] not in special]
    non_actionable = [w for w in scored_workflows if w["result"]["tier"] in special]
    actionable.sort(key=lambda w: w["result"]["score"], reverse=True)
    return actionable + non_actionable


def estimated_hours_saved(scored_workflows: list[dict]) -> float:
    """
    Estimate monthly hours recoverable from Automate Now + Automate Soon workflows.
    Uses midpoint of time cost range * 4 weeks.
    """
    from config import TIME_COST_OPTIONS
    MIDPOINTS = {
        "Less than 1 hr/week":  0.5,
        "1–2 hrs/week":         1.5,
        "2–5 hrs/week":         3.5,
        "5–10 hrs/week":        7.5,
        "10–20 hrs/week":      15.0,
        "20+ hrs/week":        25.0,
    }
    actionable_tiers = {TIER_AUTOMATE_NOW, TIER_AUTOMATE_SOON}
    total = 0.0
    for w in scored_workflows:
        if w["result"]["tier"] in actionable_tiers:
            hrs_week = MIDPOINTS.get(w["time_cost"], 0)
            total += hrs_week * 4  # weekly → monthly
    return round(total, 1)

File: src/app/tools/audit/scoring.ts
/**
 * Core scoring engine. All business logic — no React imports.
 */

import {
  WEIGHTS,
  TIME_COST_OPTIONS,
  ERROR_RATE_OPTIONS,
  DEPENDENCY_RISK_OPTIONS,
  TIER_CONFIG,
  TIER_AUTOMATE_NOW,
  TIER_AUTOMATE_SOON,
  TIER_EVALUATE,
  TIER_LOW_PRIORITY,
  TIER_DO_NOT_AUTOMATE,
  TIER_CONDITIONAL,
  JUDGMENT_OVERRIDE_THRESHOLD,
  type Tier,
} from "./config";

// -- Types -------------------------------------------------------------------

export interface WorkflowInput {
  name: string;
  description: string;
  time_cost: string;
  error_rate: string;
  dependency_risk: string;
  data_quality: string;
  judgment_pct: number;
}

export interface ScoreBreakdown {
  time_score: number;
  error_score: number;
  judgment_inverse: number;
  dependency_score: number;
}

export interface ScoreResult {
  score: number;
  tier: Tier;
  flag: string | null;
  breakdown: ScoreBreakdown;
}

export interface ScoredWorkflow extends WorkflowInput {
  result: ScoreResult;
}

export interface OverallReadiness {
  overall_score: number;
  tier_counts: Record<Tier, number>;
  automate_now: number;
  do_not_automate: number;
  conditional: number;
  actionable_count: number;
  total_count: number;
}

// -- Normalizers -------------------------------------------------------------

function normalizeTime(label: string): number {
  return TIME_COST_OPTIONS[label] ?? 0;
}

function normalizeError(label: string): number {
  return ERROR_RATE_OPTIONS[label] ?? 0;
}

function normalizeDependency(label: string): number {
  return DEPENDENCY_RISK_OPTIONS[label] ?? 0;
}

function scoreToTier(score: number): Tier {
  for (const tier of [
    TIER_AUTOMATE_NOW,
    TIER_AUTOMATE_SOON,
    TIER_EVALUATE,
    TIER_LOW_PRIORITY,
  ] as const) {
    if (score >= TIER_CONFIG[tier].min) return tier;
  }
  return TIER_LOW_PRIORITY;
}

// -- Core scoring -----------------------------------------------------------

export function calculatePriorityScore(
  workflow: Pick<
    WorkflowInput,
    "time_cost" | "error_rate" | "dependency_risk" | "data_quality" | "judgment_pct"
  >
): ScoreResult {
  const time_score = normalizeTime(workflow.time_cost);
  const error_score = normalizeError(workflow.error_rate);
  const judgment_inverse = 100.0 - workflow.judgment_pct;
  const dep_score = normalizeDependency(workflow.dependency_risk);

  const base_score = Math.round(
    (time_score * WEIGHTS.time_cost +
      error_score * WEIGHTS.error_rate +
      judgment_inverse * WEIGHTS.judgment_inverse +
      dep_score * WEIGHTS.dependency_risk) *
      10
  ) / 10;

  const breakdown: ScoreBreakdown = {
    time_score,
    error_score,
    judgment_inverse,
    dependency_score: dep_score,
  };

  // Data quality flag overrides tier (but not score)
  if (workflow.data_quality.startsWith("Poor")) {
    return {
      score: base_score,
      tier: TIER_CONDITIONAL,
      flag: "Fix data quality first — automate once inputs are structured",
      breakdown,
    };
  }

  // Judgment override
  if (workflow.judgment_pct > JUDGMENT_OVERRIDE_THRESHOLD) {
    return {
      score: base_score,
      tier: TIER_DO_NOT_AUTOMATE,
      flag: "Human judgment is the value — automation would degrade quality",
      breakdown,
    };
  }

  return {
    score: base_score,
    tier: scoreToTier(base_score),
    flag: null,
    breakdown,
  };
}

// -- Aggregates -------------------------------------------------------------

export function calculateOverallReadiness(
  scored: ScoredWorkflow[]
): OverallReadiness {
  const actionable = scored.filter(
    (w) =>
      w.result.tier !== TIER_DO_NOT_AUTOMATE &&
      w.result.tier !== TIER_CONDITIONAL
  );

  const avg =
    actionable.length === 0
      ? 0
      : actionable.reduce((sum, w) => sum + w.result.score, 0) /
        actionable.length;

  const tier_counts = Object.fromEntries(
    Object.keys(TIER_CONFIG).map((t) => [t, 0])
  ) as Record<Tier, number>;

  for (const w of scored) {
    tier_counts[w.result.tier] = (tier_counts[w.result.tier] || 0) + 1;
  }

  return {
    overall_score: Math.round(avg * 10) / 10,
    tier_counts,
    automate_now: tier_counts[TIER_AUTOMATE_NOW] || 0,
    do_not_automate: tier_counts[TIER_DO_NOT_AUTOMATE] || 0,
    conditional: tier_counts[TIER_CONDITIONAL] || 0,
    actionable_count: actionable.length,
    total_count: scored.length,
  };
}

export function rankWorkflows(scored: ScoredWorkflow[]): ScoredWorkflow[] {
  const special = [TIER_DO_NOT_AUTOMATE, TIER_CONDITIONAL] as Tier[];
  const actionable = scored.filter((w) => !special.includes(w.result.tier));
  const nonActionable = scored.filter((w) => special.includes(w.result.tier));
  actionable.sort((a, b) => b.result.score - a.result.score);
  return [...actionable, ...nonActionable];
}

const TIME_MIDPOINTS: Record<string, number> = {
  "Less than 1 hr/week": 0.5,
  "1–2 hrs/week": 1.5,
  "2–5 hrs/week": 3.5,
  "5–10 hrs/week": 7.5,
  "10–20 hrs/week": 15.0,
  "20+ hrs/week": 25.0,
};

export function estimatedHoursSaved(scored: ScoredWorkflow[]): number {
  const actionableTiers = new Set<Tier>([TIER_AUTOMATE_NOW, TIER_AUTOMATE_SOON]);
  let total = 0;
  for (const w of scored) {
    if (actionableTiers.has(w.result.tier)) {
      const hrs = TIME_MIDPOINTS[w.time_cost] ?? 0;
      total += hrs * 4; // weekly -> monthly
    }
  }
  return Math.round(total * 10) / 10;
}
